commands/upgrade_cluster.go

```diff
@@ -0,0 +1,344 @@
+	default:
+		headline = err.Error()
+	}
+
+	// print output
+	fmt.Println(color.RedString(headline))
+	if subtext != "" {
+		fmt.Println(subtext)
+	}
+	os.Exit(1)
+}
+
+// Checks if all preconditions are met, before actually executing
+// our business function
+func verifyUpgradeClusterPreconditions(args upgradeClusterArguments, cmdLineArgs []string) error {
+	// authentication
+	if config.Config.Token == "" && args.authToken == "" {
+		return microerror.Mask(notLoggedInError)
+	}
+
+	// cluster ID is present
+	if args.clusterID == "" {
+		return microerror.Mask(clusterIDMissingError)
+	}
+
+	return nil
+}
+
+// upgradeClusterRunOutput executes our business function and displays the result,
+// both in case of success or error
+func upgradeClusterRunOutput(cmd *cobra.Command, cmdLineArgs []string) {
+	args := defaultUpgradeClusterArguments(cmdLineArgs)
+	result, err := upgradeCluster(args)
+
+	if err != nil {
+		var headline = ""
+		var subtext = ""
+
+		switch {
+		case err.Error() == "":
+			return
+		case IsCouldNotCreateClientError(err):
+			headline = "Failed to create API client."
+			subtext = "Details: " + err.Error()
+		case IsNoUpgradeAvailableError(err):
+			headline = "There is no newer release available."
+			subtext = "Please check the available releases using 'gsctl list releases'."
+		case IsClusterNotFoundError(err):
+			headline = "The cluster does not exist."
+			subtext = fmt.Sprintf("We couldn't find a cluster with the ID '%s' via API endpoint %s.", args.clusterID, args.apiEndpoint)
+		case IsCommandAbortedError(err):
+			headline = "Not upgrading."
+		default:
+			headline = err.Error()
+		}
+
+		// Print error output
+		fmt.Println(color.RedString(headline))
+		if subtext != "" {
+			fmt.Println(subtext)
+		}
+		os.Exit(1)
+	}
+
+	fmt.Println(color.GreenString("Starting to upgrade cluster '%s' to release version %s",
+		color.CyanString(result.clusterID),
+		color.CyanString(result.versionAfter)))
+}
+
+// upgradeCluster performs our actual function. It usually creates an API client,
+// configures it, configures an API request and performs it.
+func upgradeCluster(args upgradeClusterArguments) (upgradeClusterResult, error) {
+	result := upgradeClusterResult{}
+
+	// fetch current cluster details
+	details, detailsErr := getClusterDetails(args.clusterID, args.authToken, args.apiEndpoint)
+	if detailsErr != nil {
+		return result, microerror.Mask(detailsErr)
+	}
+
+	listReleasesArgs := listReleasesArguments{
+		apiEndpoint: args.apiEndpoint,
+		token:       args.authToken,
+	}
+	releasesResult, releasesErr := listReleases(listReleasesArgs)
+	if releasesErr != nil {
+		return result, microerror.Mask(releasesErr)
+	}
+	releaseVersions := []string{}
+	for _, r := range releasesResult.releases {
+		releaseVersions = append(releaseVersions, r.Version)
+	}
+
+	// define the target version to upgrade to
+	targetVersion := successorReleaseVersion(details.ReleaseVersion, releaseVersions)
```

Comment:
Some of our older clusters have no release version because they were made before we had that feature.

Those clusters should also be considered not upgradeable I think. 

Not totally sure what this code will do in case `details.ReleaseVersion` == `""`

Reply:
Sorry to have ignored this old comment. Not sure why... I have added treatment for this case to `successorReleaseVersion()`.